src/deployment/mod.rs

```diff
@@ -0,0 +1,1086 @@
+            Err(err) => {
+                return Err(format!(
+                    "unable to read contract at path {:?}: {}",
+                    config.path, err
+                ))
+            }
+        };
+
+        let contract = EmulatedContractPublishSpecification {
+            contract_name,
+            emulated_sender,
+            source,
+            relative_path: config.path.clone(),
+        };
+
+        let contract_id = QualifiedContractIdentifier::new(
+            contract.emulated_sender.clone(),
+            contract.contract_name.clone(),
+        );
+
+        contracts.insert(contract_id, contract);
+    }
+
+    let session = Session::new(settings);
+
+    let mut contract_asts = HashMap::new();
+    let mut contract_diags = HashMap::new();
+
+    for (contract_id, contract) in contracts.iter() {
+        let (ast, diags, _) = session.interpreter.build_ast(
+            contract_id.clone(),
+            contract.source.clone(),
+            parser_version,
+        );
+        contract_asts.insert(contract_id.clone(), ast);
+        contract_diags.insert(contract_id.clone(), diags);
+    }
+
+    let dependencies =
+        ASTDependencyDetector::detect_dependencies(&contract_asts, &requirements_asts);
+
+    let mut dependencies = match dependencies {
+        Ok(dependencies) => dependencies,
+        Err(_) => {
+            return Err(format!("unable to detect dependencies"));
+        }
+    };
+
+    for contract_id in boot_contracts_ids.into_iter() {
+        dependencies.insert(contract_id.clone(), DependencySet::new());
+    }
+
+    dependencies.extend(requirements_deps);
+
+    let ordered_contracts_ids = match ASTDependencyDetector::order_contracts(&dependencies) {
+        Ok(ordered_contracts_ids) => ordered_contracts_ids
+            .into_iter()
+            .map(|c| c.clone())
+            .collect::<Vec<_>>(),
+        Err(e) => return Err(format!("unable to order contracts {}", e)),
+    };
+
+    for contract_id in ordered_contracts_ids.into_iter() {
+        if requirements_asts.contains_key(&contract_id) {
+            continue;
+        }
+        let data = contracts
+            .remove(&contract_id)
+            .expect("unable to retrieve contract");
+        contracts_map.insert(
+            contract_id.clone(),
+            (data.source.clone(), data.relative_path.clone()),
+        );
+        let tx = if network.is_none() {
+            TransactionSpecification::EmulatedContractPublish(data)
+        } else {
+            TransactionSpecification::ContractPublish(ContractPublishSpecification {
+                contract_name: data.contract_name.clone(),
+                expected_sender: data.emulated_sender.clone(),
+                relative_path: data.relative_path.clone(),
+                source: data.source.clone(),
+                cost: deployment_fee_rate
+                    .saturating_mul(data.source.as_bytes().len().try_into().unwrap()),
+            })
+        };
```

Comment:
This feels a little strange creating a `ContractPublish` from an `EmulatedContractPublish`. Why not just create the right type originally when they are added to `contracts` above?

Reply:
Addressed with latest